### ISSUE TITLE
Thread termination

### ISSUE THREAD

**Issue by paulfloyd** [paulfloyd/macos_valgrind#4]
Not working properly resulting in panic

```
      mach_msg_header_t msg;

      VG_(debugLog)(1, "syswrap-darwin",
                       "run_a_thread_NORETURN(tid=%u): "
                          "not last one standing\n", 
                          tid); 

      /* OK, thread is dead, but others still exist.  Just exit. */
   
      /* This releases the run lock */
      VG_(exit_thread)(tid);
      vg_assert(tst->status == VgTs_Zombie);
   
      /* tid is now invalid. */

      // GrP fixme exit race
      msg.msgh_bits = MACH_MSGH_BITS(17, MACH_MSG_TYPE_MAKE_SEND_ONCE);
      msg.msgh_request_port = VG_(gettid)();
      msg.msgh_reply_port = 0;
      msg.msgh_id = 3600;  // thread_terminate

      tst->status = VgTs_Empty;
      // GrP fixme race here! new thread may claim this V thread stack
      // before we get out here!
      // GrP fixme use bsdthread_terminate for safe cleanup?
      mach_msg(&msg, MACH_SEND_MSG|MACH_MSG_OPTION_NONE,
               sizeof(msg), 0, 0, MACH_MSG_TIMEOUT_NONE, 0);

      // DDD: This is reached sometimes on none/tests/manythreads, maybe
      // because of the race above.
      VG_(core_panic)("Thread exit failed?\n");
```

I tried moving the msg stuff before the VG_(exit_thread) to try to reduce the chances of the race.

It looks like the MACH_SEND_MSG isn't working (or has changed).

Stepping through the exit code in assembler on macOS 13.2 I do see that bsdthread_terminate gets called.

I'll try to see if I can do the same for my old 10.7 machine and 10.13 / 10.14 vbox machines.

bsdthread_terminate does look a bit tricky
361	AUE_NULL	ALL	{ int bsdthread_terminate(user_addr_t stackaddr, size_t freesize, uint32_t port, uint32_t sem) NO_SYSCALL_STUB; } 

from

https://github.com/opensource-apple/xnu/blob/master/bsd/kern/syscalls.master

Looking at

https://github.com/apple-oss-distributions/libpthread/blob/libpthread-514/src/pthread.c

```
_pthread_exit(pthread_t self, void *exit_value)
{
	struct __darwin_pthread_handler_rec *handler;

	// Disable signal delivery while we clean up
	__disable_threadsignal(1);

	// Set cancel state to disable and type to deferred
	_pthread_setcancelstate_exit(self, exit_value);

	while ((handler = self->__cleanup_stack) != 0) {
		(handler->__routine)(handler->__arg);
		self->__cleanup_stack = handler->__next;
	}
	_pthread_tsd_cleanup(self);

	// Clear per-thread semaphore cache
	os_put_cached_semaphore(SEMAPHORE_NULL);

	_pthread_terminate_invoke(self, exit_value);
}
```

_pthread_terminate_invoke dies some thread stack protection and calls _pthread_terminate

That does a lot of stuff.

What does __bsdthread_terminate need for its arguments? If I can figure that out I can try fairly quicky to use it.

Otherwise there's a fair bit of work to get to understand what macOS does for pthread_create and exit and what correspongling Valgrind is doing.

In short it seems that we intercept bsdthread_register / bsdthread_create and instead of using the client pthread function we use pthread_hijack which runs the client function on the host.

run_a_thread_NORETURN

**Comment by paulfloyd** [comment]
If we use bsdthread_terminate that need 4 arguments

1. freeaddr

this is a field of pthread_t

2. freesize

and so is this but it may get modified

3. kport

mach_port_t kport = _pthread_tsd_slot(t, MACH_THREAD_SELF);

4. custom_stack_sema

	if (t->tl_join_ctx) {
		custom_stack_sema = _pthread_joiner_prepost_wake(t); // unsets tl_joinable
		needs_wake = true;
	}

will pthread_join work without this?

3 questions here

- Can I get the pthread_t pointer? it does get passed to bsdthread_create so we could store it
- where is _pthread_tsd_slot ?
- can we do the _pthread_joiner_prepost_wake stuff?

Answering the second question

#define _pthread_tsd_slot(th, name) \
                (*(_PTHREAD_TSD_SLOT_##name##_TYPE *)(uintptr_t *)&(th)->tsd[_PTHREAD_TSD_SLOT_##name])

And the third

semaphore_t
_pthread_joiner_prepost_wake(pthread_t thread)
{
        pthread_join_context_t ctx = thread->tl_join_ctx;
        semaphore_t sema = MACH_PORT_NULL;
 
        if (thread->tl_joinable) {
                sema = ctx->custom_stack_sema;
                thread->tl_joinable = false;
        } else {
                ctx->detached = true;
                thread->tl_join_ctx = NULL;
        }
        if (ctx->value_ptr) *ctx->value_ptr = _pthread_get_exit_value(thread);
        return sema;
}



So I think I'll try to save the pthread_t pointer and we should be able to do all of the rest

**Comment by paulfloyd** [comment]
Seems fixed by Louis.
